ocaml/xapi/rpm.ml

```diff
@@ -0,0 +1,290 @@
+(*
+ * Copyright (C) Citrix Systems Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation; version 2.1 only. with the special
+ * exception on linking described in file LICENSE.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details.
+ *)
+
+module D = Debug.Make (struct let name = "rpm" end)
+
+open D
+
+module Epoch = struct
+  type t = int option
+
+  let epoch_none = "(none)"
+
+  exception Invalid_epoch
+
+  let of_string = function
+    | s when s = epoch_none || s = "None" ->
+        None
+    | s -> (
+      match int_of_string s with
+      | i when i = 0 ->
+          None
+      | i when i > 0 ->
+          Some i
+      | _ ->
+          raise Invalid_epoch
+      | exception _ ->
+          raise Invalid_epoch
+    )
+
+  let to_string = function Some i -> string_of_int i | None -> epoch_none
+end
+
+module Pkg = struct
+  type t = {
+      name: string
+    ; epoch: Epoch.t
+    ; version: string
+    ; release: string
+    ; arch: string
+  }
+
+  type order = LT | EQ | GT
+
+  type segment_of_version = Int of int | Str of string
+
+  let string_of_order = function LT -> "<" | EQ -> "=" | GT -> ">"
+
+  let error_msg = Printf.sprintf "Failed to parse '%s'"
+
+  let parse_epoch_version_release epoch_ver_rel =
```

Comment:
This module looks too complicated to me. I would expect that a version is represented as a list of numbers and that we are comparing two strings to establish their order. I think it needs to be spelled out how strings and numbers are treated: is every string that can be read as a number considered a number, and or still a string? This makes a difference because "01" and "1" are different strings but the same number. Likewise, how is any string that cannot be read as a number treated? I am not sure but I would have started with:
```
type comp = Number of int | String of string
type version = comp list (* assuming a version is a list of components, each either a string or a number *)
val parse: string -> string * version (* split into package and version *)
val compare: version -> version -> int (* -1, 0, 1 *)
```

Reply:
These bits are some existing lines which are moved into a new file only in this PR.

`type comp = Number of int | String of string`
This is defined as similarly as:
https://github.com/xapi-project/xen-api/pull/4678/files#diff-3c831fd11744c8735ff48fa1f9e436f357cb4316e4394855770ac300ed41dc69R55

`val parse: string -> string * version (* split into package and version *)`
is implemented like:
https://github.com/xapi-project/xen-api/pull/4678/files#diff-3c831fd11744c8735ff48fa1f9e436f357cb4316e4394855770ac300ed41dc69R183

and `val compare: version -> version -> int (* -1, 0, 1 *)` is implemented as:
https://github.com/xapi-project/xen-api/pull/4678/files#diff-3c831fd11744c8735ff48fa1f9e436f357cb4316e4394855770ac300ed41dc69R160

There are some examples to explain the rules of comparison:
https://github.com/xapi-project/xen-api/pull/4678/files#diff-3c831fd11744c8735ff48fa1f9e436f357cb4316e4394855770ac300ed41dc69R166